File: EXAMPLES/Java/CS50 Java/pset2/Java/Substitution.java
import java.util.Scanner;
import java.util.HashMap;

/*
* Testing:
* 
* java Substitution YTNSHKVEFXRBAUQZCLWDMIPGJO
* plaintext:  HELLO
* ciphertext: EHBBQ
* 
* java Substitution VCHPRZGJNTLSKFBDQWAXEUYMOI
* plaintext:  hello, world
* ciphertext: jrssb, ybwsp
* 
*/


public class Substitution {

    public static int KEY_LENGTH = 26;
    public static Scanner myScanner = new Scanner(System.in);

    public static Boolean validateKey(String key) {

        // HashMap with 26 Booleans to remember found characters by their ASCII codes
         HashMap<Integer, Boolean> foundCharacters = new HashMap<Integer, Boolean>();
         for (int i = 0; i < KEY_LENGTH; i++) {
            foundCharacters.put(i+97, false);
         }

         int uniqueCharacters = 0;

         for (int i = 0, l = key.length(); i < l; i++) {
            
            char currentCharacter = Character.toLowerCase(key.charAt(i));

            // Invalid characters (non alphabetic)
            if (!Character.isAlphabetic(currentCharacter)) {
                // System.out.println("Invalid 1");
                return false;
            }
            
            // Repeated characters (already in foundCharacters HashMap)
            if (foundCharacters.get((int) currentCharacter)) {
                // System.out.println("Invalid 2");
                return false;
            } else {
                foundCharacters.put((int) currentCharacter, true);
                uniqueCharacters++;
            }
            
        }

        // 26 unique characters
        if (uniqueCharacters != KEY_LENGTH) {
            // System.out.println("Invalid 3");
            return false;
        }

        return true;
    }

    public static void main(String[] args) {

        // Correct number of arguments
        if (args.length != 1) {
            System.out.println("Usage: java Substitution key");
            System.exit(1);
        }

        // Key is valid (26 unique alphabetic)
        if (!validateKey(args[0])) {
            System.out.println("Invalid key");
            System.exit(1);
        }

        // Stores key in a variable
        String key = args[0];

        // Gets plaintext from user input
        System.out.printf("plaintext: ");
        String plaintext = myScanner.nextLine();

        // Print ciphertext
        // Use ASCII values of the plaintext chars to index the key        
        System.out.printf("ciphertext: ");
        for (int i = 0, l = plaintext.length(); i < l; i++) {
            if (Character.isLowerCase(plaintext.charAt(i))) {
                int cipherIndex = Character.toLowerCase(plaintext.charAt(i))-97;
                System.out.print(Character.toLowerCase(key.charAt(cipherIndex)));
            } else if (Character.isUpperCase(plaintext.charAt(i))) {
                int cipherIndex = Character.toLowerCase(plaintext.charAt(i))-97;
                System.out.print(Character.toUpperCase(key.charAt(cipherIndex)));
            } else {
                System.out.print(plaintext.charAt(i));
            }
         }
         System.out.printf("\n");
    }

}

File: EXAMPLES/Java/CS50 Java/pset2/C/substitution.c
#include <cs50.h>
#include <stdio.h>
#include <stdbool.h>
#include <ctype.h>
#include <string.h>

/*
* Link cs50.h when compiling
* gcc substitution.c -o substitution -lcs50
* 
* Use check50 locally
* check50 --local cs50/problems/2022/x/substitution
*
* Testing:
* 
* ./substitution YTNSHKVEFXRBAUQZCLWDMIPGJO
* plaintext:  HELLO
* ciphertext: EHBBQ
* 
* ./substitution VCHPRZGJNTLSKFBDQWAXEUYMOI
* plaintext:  hello, world
* ciphertext: jrssb, ybwsp
* 
*/


const int KEY_LENGHT = 26;
bool validate_key(string key);

int main(int argc, string argv[])
{
    // Correct number of arguments
    if (argc != 2)
    {
        printf("Usage: ./substitution key\n");
        return 1;
    }

    // Key is valid (26 letters, all unique)
    if (!validate_key(argv[1]))
    {
        printf("Invalid key\n");
        return 1;
    }

    // Stores key in a variable
    char key[KEY_LENGHT];
    strcpy(key, argv[1]);

    // Gets plaintext from user input - ch50.h library function
    string plaintext = get_string("plaintext: ");

    // Print ciphertext

    /*
    * We use the ASCII values of the plaintext characters
    * to index the key, getting each ciphertext character.
    * 
    */

    printf("ciphertext: ");
    for (int i = 0, l = strlen(plaintext); i < l; i++)
    {
        if (islower(plaintext[i]))
        {
            int cipher_index = tolower(plaintext[i]) - 'a';
            printf("%c", tolower(key[cipher_index]));
        }
        else if (isupper(plaintext[i]))
        {
            int cipher_index = tolower(plaintext[i]) - 'a';
            printf("%c", toupper(key[cipher_index]));
        }
        else
        {
            printf("%c", plaintext[i]);
        }
    }
    printf("\n");
}


bool validate_key(string key)
{
    /*
    * ARRAY OF 26 BOOLEANS TO REMEMBER FOUND CHARACTERS
    * Uses ASCII value to index the array and set boolean to true,
    * so later we will know if the character was found before.
    */

    bool all_characters[KEY_LENGHT];
    for (int i = 0, l = KEY_LENGHT; i < l; i++)
    {
        all_characters[i] = false;
    }

    int unique_characters = 0;

    for (int i = 0, l = strlen(key); i < l; i++)
    {
        int current_character = tolower(key[i]);

        // Invalid characters (non letters)
        if (current_character < 'a' || current_character > 'z')
        {
            return false;
        }

        // Repeated characters (already found before)
        if (all_characters[current_character - 'a'] == true) {
            return false;
        }
        else
        {
            all_characters[current_character - 'a'] = true;
            unique_characters++;
        }
    }

    // Must have 26 unique characters
    if (unique_characters != KEY_LENGHT)
    {
        return false;
    }

    return true;
}
